File: translate_missing.js
import fs from 'fs';
import translate from 'translate';

// Use a free engine
translate.engine = 'google';

const file = './src/data/verbsData.json';
const data = JSON.parse(fs.readFileSync(file, 'utf8'));

async function translateAll() {
  let count = 0;
  for (let i = 0; i < data.roots.length; i++) {
    const r = data.roots[i];
    const m = r.meaning;
    // Condition to check if it's missing (just Buckwalter)
    if (!m.includes(' ') && !m.startsWith('to ')) {
      try {
        // Translate the Arabic root word directly
        const arabic = r.root;
        const res = await translate(arabic, { from: 'ar', to: 'en' });
        // Format as "to <verb>" or something similar
        r.meaning = res.toLowerCase();
        count++;
        console.log(`Translated [${arabic}] -> ${r.meaning}`);
        // Small delay to avoid rate limiting
        await new Promise(resolve => setTimeout(resolve, 300));
      } catch (err) {
        console.error(`Failed on ${r.root}:`, err.message);
      }
    }
  }

  if (count > 0) {
    fs.writeFileSync(file, JSON.stringify(data, null, 2));
    console.log(`Updated ${count} missing meanings.`);
  } else {
    console.log('No missing meanings found.');
  }
}

translateAll();

File: translate_missing.py
import json
import time
import sys
from deep_translator import GoogleTranslator

def main():
    file_path = 'src/data/verbsData.json'
    
    print("Loading data...")
    with open(file_path, 'r', encoding='utf-8') as f:
        data = json.load(f)

    # Initialize translator from Arabic to English
    translator = GoogleTranslator(source='ar', target='en')

    count = 0
    total_missing = 0
    
    # Pre-calculate to show progress
    for r in data['roots']:
        m = r.get('meaning', '')
        if not (' ' in m or m.startswith('to ')):
            total_missing += 1

    print(f"Found {total_missing} roots needing translation.")

    if total_missing == 0:
        return

    # Process each root
    for r in data['roots']:
        m = r.get('meaning', '')
        
        # If the meaning doesn't have a space and doesn't start with "to ", we assume it's just the buckwalter code.
        if not (' ' in m or m.startswith('to ')):
            root_arabic = r['root']
            try:
                # deep_translator throws an error or returns the input if not found, we use it directly:
                en = translator.translate(root_arabic)
                
                # Simple cleanup (lowercasing and prepending 'to' if it's not there and seems like a single verb translation)
                en = en.lower()
                if ' ' not in en and not en.startswith('to '):
                    en = f"to {en}"

                print(f"[{count+1}/{total_missing}] Translated {root_arabic} ({m}) -> {en}")
                r['meaning'] = en
                count += 1
                
                # Small sleep to be polite to the Translation API
                time.sleep(0.4)
                
            except Exception as e:
                print(f"Error translating {root_arabic}: {e}")
                time.sleep(1) # wait longer on error

            # Save periodically to not lose progress on failure
            if count % 20 == 0:
                with open(file_path, 'w', encoding='utf-8') as f:
                    json.dump(data, f, ensure_ascii=False, indent=2)

    # Final save
    if count > 0:
        with open(file_path, 'w', encoding='utf-8') as f:
            json.dump(data, f, ensure_ascii=False, indent=2)
        print(f"\nDone! Successfully updated {count} meanings.")

if __name__ == '__main__':
    main()
